File: py/src/nearai_inference_sdk/utils/sse.py
"""SSE record parsing shared by decryption and completion receipts."""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class SseLine:
    content: str
    ending: str


@dataclass(frozen=True)
class SseRecord:
    value: str
    separator: str


@dataclass(frozen=True)
class CompleteSseRecords:
    records: list[SseRecord]
    pending: str


def take_complete_sse_records(value: str) -> CompleteSseRecords:
    """Split complete events, keeping a trailing CR for the next byte chunk."""

    records: list[SseRecord] = []
    record_start = line_start = index = 0
    while index < len(value):
        if value[index] not in '\r\n':
            index += 1
            continue
        if value[index] == '\r' and index + 1 == len(value):
            break
        line_end = index
        index += 2 if value[index : index + 2] == '\r\n' else 1
        if line_end == line_start:
            records.append(
                SseRecord(value[record_start:line_start], value[line_start:index])
            )
            record_start = index
        line_start = index
    return CompleteSseRecords(records, value[record_start:])


def split_sse_lines(value: str) -> list[SseLine]:
    lines: list[SseLine] = []
    start = index = 0
    while index < len(value):
        if value[index] not in '\r\n':
            index += 1
            continue
        ending = '\r\n' if value[index : index + 2] == '\r\n' else value[index]
        lines.append(SseLine(value[start:index], ending))
        index += len(ending)
        start = index
    if start < len(value):
        lines.append(SseLine(value[start:], ''))
    return lines


def get_sse_data(line: str) -> str | None:
    if line == 'data':
        return ''
    if not line.startswith('data:'):
        return None
    value = line[len('data:') :]
    return value[1:] if value.startswith(' ') else value


def get_sse_data_records(text: str) -> list[str]:
    """Read complete response data payloads, joining each event's data lines."""

    complete = take_complete_sse_records(text)
    values = [record.value for record in complete.records]
    if complete.pending:
        values.append(complete.pending)
    return [
        '\n'.join(
            data
            for line in split_sse_lines(value)
            if (data := get_sse_data(line.content)) is not None
        )
        for value in values
    ]

File: js/src/utils/sse.ts
export type SseLine = {
  readonly content: string;
  readonly ending: string;
};

type CompleteSseRecord = {
  /** Event lines, including the final line ending. */
  readonly value: string;
  /** The blank line ending that terminates the event. */
  readonly separator: string;
};

type CompleteSseRecords = {
  readonly records: readonly CompleteSseRecord[];
  readonly pending: string;
};

/** Split complete events without treating the CR and LF in CRLF as two lines. */
export function takeCompleteSseRecords(value: string): CompleteSseRecords {
  const records: CompleteSseRecord[] = [];
  let recordStart = 0;
  let lineStart = 0;
  let index = 0;
  while (index < value.length) {
    if (value[index] !== '\r' && value[index] !== '\n') {
      index += 1;
      continue;
    }
    // A trailing CR may be the first half of a CRLF split across chunks.
    if (value[index] === '\r' && index + 1 === value.length) break;
    const lineEnd = index;
    index += value[index] === '\r' && value[index + 1] === '\n' ? 2 : 1;
    if (lineEnd === lineStart) {
      records.push({
        value: value.slice(recordStart, lineStart),
        separator: value.slice(lineStart, index),
      });
      recordStart = index;
    }
    lineStart = index;
  }
  return { records, pending: value.slice(recordStart) };
}

export function splitSseLines(value: string): SseLine[] {
  const lines: SseLine[] = [];
  let start = 0;
  for (let index = 0; index < value.length; index += 1) {
    if (value[index] !== '\n' && value[index] !== '\r') continue;
    const isCrLf = value[index] === '\r' && value[index + 1] === '\n';
    const ending = isCrLf ? '\r\n' : value[index];
    lines.push({ content: value.slice(start, index), ending });
    start = index + ending.length;
    if (isCrLf) index += 1;
  }
  if (start < value.length) {
    lines.push({ content: value.slice(start), ending: '' });
  }
  return lines;
}

export function getSseData(line: string): string | undefined {
  if (line === 'data') return '';
  if (!line.startsWith('data:')) return undefined;
  const value = line.slice('data:'.length);
  return value.startsWith(' ') ? value.slice(1) : value;
}

/** Read event payloads from a complete response body, joining its data lines. */
export function getSseDataRecords(text: string): string[] {
  const { records, pending } = takeCompleteSseRecords(text);
  const values = records.map((record) => record.value);
  if (pending.length > 0) values.push(pending);
  return values.map((value) =>
    splitSseLines(value)
      .map((line) => getSseData(line.content))
      .filter((data) => data !== undefined)
      .join('\n'),
  );
}
